### PR TITLE
GH-347 remote repo password always update

### LINKED ISSUE #347: artifactory_remote_maven_repository always updating password in plan/apply
**Describe the bug**

**Requirements for and issue**
- [x] A fully functioning terraform snippet that can be copy&pasted (no outside files or ENV vars unless that's part of the issue)
- [x] Your version of artifactory (you can `curl` it at `$host/artifactory/api/system/version`
- [x] Your version of terraform

**Expected behavior**

**Additional context**

Terraform v1.1.7
on linux_amd64
- Installing hashicorp/aws v4.3.0...
- Installing hashicorp/helm v2.4.1...
- Installing hashicorp/kubernetes v2.8.0...
- Installing hashicorp/random v3.1.0...
- Installing jfrog/artifactory v2.21.0...
- Installing hashicorp/vault v3.3.1...

```hcl
resource "artifactory_remote_maven_repository" "mulesoft-ee" {
  key                            = "mulesoft-ee"
  repo_layout_ref                = "maven-2-default"
  url                            = data.vault_generic_secret.mulesoft.data["url"]
  handle_releases                = true
  handle_snapshots               = true
  includes_pattern               = "**/*"
  username                       = data.vault_generic_secret.mulesoft.data["username"]
  password                       = data.vault_generic_secret.mulesoft.data["password"]
  socket_timeout_millis          = "5000"
  store_artifacts_locally        = true
  block_mismatching_mime_types   = true
  retrieval_cache_period_seconds = "6000"
  missed_cache_period_seconds    = "1800"
  project_environments           = []
  property_sets                  = []
}
```

Terraform Plan:

Artifactory 107.33.12

Terraform will perform the following actions:
```
  # artifactory_remote_maven_repository.mulesoft-ee will be updated in-place
  ~ resource "artifactory_remote_maven_repository" "mulesoft-ee" {
        id                                    = "mulesoft-ee"
      ~ password                              = (sensitive value)
        # (43 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

After applying it will continue to state it will update in place. This happens whether password is hard coded or a data lookup. 

Terraform state shows:
```
# artifactory_remote_maven_repository.mulesoft-ee:
resource "artifactory_remote_maven_repository" "mulesoft-ee" {
    allow_any_host_auth                   = false
...
    password                              = (sensitive value)

...
   unused_artifacts_cleanup_period_hours = 0
    url                                   = (sensitive)
    username                              = (sensitive)
    xray_index                            = false
```

### PR DESCRIPTION
**Pull request by alexhung**
Fixes #347

- Remove hashing of remote repos ‘password’ attribute
- Remove ‘omitempty’ tag for the same
- Add `noPassword` predicate and `noPasswordPacker` packer so the value from Artifactory API is ignored for TF state.

### CODE REVIEW

**Comment by alexhung** [comment]
This doesn't conflict or change #380 so we can merge this before #380.

**Review by danielmkn — APPROVED** [review]

**Review by maheshjfrog — APPROVED** [review]

**Review by chb0github — DISMISSED** [review]
No real requests for changes, just a suggestion. I would love to understand this issue a little better

**Review by dasmanas — DISMISSED** [review]

File: pkg/artifactory/resource_artifactory_remote_cocoapods_repository.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ func resourceArtifactoryRemoteCocoapodsRepository() *schema.Resource {
 		return repo, repo.Id(), nil
 	}
 
-	return mkResourceSchema(cocoapodsRemoteSchema, defaultPacker, unpackCocoapodsRemoteRepo, func() interface{} {
+	return mkResourceSchema(cocoapodsRemoteSchema, noPasswordPacker, unpackCocoapodsRemoteRepo, func() interface{} {
```
[the remark this answers is not in the document]
Comment (alexhung) [reply pkg/artifactory/resource_artifactory_remote_cocoapods_repository.go:37]: Yea, I see the other way of composing the packer with `inSchema` but I wasn't sure I need to here (as well as not wanting to make too big changes).

**Comment by alexhung** [comment]
@chb0github Replaced `noPasswordPacker` with your suggestion.

File: pkg/artifactory/resource_artifactory_remote_go_repository.go [hunk 2]
```diff
@@ -32,7 +32,14 @@ func resourceArtifactoryRemoteGoRepository() *schema.Resource {
 		return repo, repo.Id(), nil
 	}
 
-	return mkResourceSchema(goRemoteSchema, defaultPacker, unpackGoRemoteRepo, func() interface{} {
+	goRemoteRepoPacker := universalPack(
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_remote_go_repository.go:35]: How about a default packer that takes an argument that is the schema?

**Review by chb0github — CHANGES_REQUESTED** [review]
Too much CP

I think what you've done could replace the default packer

File: pkg/artifactory/repositories.go [hunk 3]
```diff
@@ -1275,25 +1272,26 @@ func composePacker(packers ...PackFunc) PackFunc {
...
-func inSchema(skeema map[string]*schema.Schema) func(payload interface{}, d *schema.ResourceData) error {
-	return universalPack(schemaHasKey(skeema))
-}
+var defaultPacker = universalPack(nil, noClass)
 
 // universalPack consider making this a function that takes a predicate of what to include and returns
 // a function that does the job. This would allow for the legacy code to specify which keys to keep and not
-func universalPack(predicate HclPredicate) func(payload interface{}, d *schema.ResourceData) error {
+func universalPack(skeema map[string]*schema.Schema, predicates ...HclPredicate) func(payload interface{}, d *schema.ResourceData) error {
+	if skeema != nil && len(skeema) > 0 {
```
Comment (chb0github) [inline pkg/artifactory/repositories.go:1280]: I think `len(skeema)` handles nil

Comment (alexhung) [reply pkg/artifactory/repositories.go:1280]: Unfortunately, it doesn't.

File: pkg/artifactory/repositories.go [hunk 4]
```diff
@@ -1275,25 +1272,26 @@ func composePacker(packers ...PackFunc) PackFunc {
 	}
 }
 
-var defaultPacker = universalPack(noClass)
-
-func inSchema(skeema map[string]*schema.Schema) func(payload interface{}, d *schema.ResourceData) error {
-	return universalPack(schemaHasKey(skeema))
-}
+var defaultPacker = universalPack(nil, noClass)
```
Comment (chb0github) [inline pkg/artifactory/repositories.go:1275]: I was thinking something more like:

```go
func defaultPacker(skeema map[string]*schema) {
   return univeralPack(hasKey(skeema,noPassword))
}
```

And then not have this code in the universalPack - I think it's unwise to cook that in
}

**Review by chb0github — APPROVED** [review]

**Pull request merged**